### PR TITLE
feat: kubo 0.29 and native apple silicon

### LINKED ISSUE #2335: [gui error report] Error: Command failed: powershell.exe -NoProfile -NonInteractive -InputF
👉️ Please describe what you were doing when this error happened.

**Specifications**

- **OS**: win32 10.0.22000
- **IPFS Desktop Version**: 0.24.0
- **Electron Version**: 19.0.9
- **Chrome Version**: 102.0.5005.167

**Error**

```
Error: Command failed: powershell.exe -NoProfile -NonInteractive -InputFormat None -Command Get-AuthenticodeSignature -LiteralPath 'C:\Users\Jknap\AppData\Local\ipfs-desktop-updater\pending\temp-IPFS-Desktop-Setup-0.24.1.exe' | ConvertTo-Json -Compress | ForEach-Object { [Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($_)) }
Get-AuthenticodeSignature : File 
C:\Users\Jknap\AppData\Local\ipfs-desktop-updater\pending\temp-IPFS-Desktop-Setup-0.24.1.exe was not found.
At line:1 char:1
  Get-AuthenticodeSignature -LiteralPath 'C:\Users\Jknap\AppData\Local\ ...
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      CategoryInfo          : ObjectNotFound: (:) [Get-AuthenticodeSignature], FileNotFoundException
      FullyQualifiedErrorId : SignatureCommandsBaseFileNotFound,Microsoft.PowerShell.Commands.GetAuthenticodeSignature 
   Command
 

    at ChildProcess.exithandler (node:child_process:408:12)
    at ChildProcess.emit (node:events:526:28)
    at maybeClose (node:internal/child_process:1092:16)
    at ChildProcess._handle.onexit (node:internal/child_process:302:5)
```

### LINKED ISSUE #2636: [gui error report] Error: Command failed: powershell.exe -NoProfile -NonInteractive -InputF
👉️ Please describe what you were doing when this error happened.

**Specifications**

- **OS**: win32 10.0.19045
- **IPFS Desktop Version**: 0.29.0
- **Electron Version**: 19.1.9
- **Chrome Version**: 102.0.5005.167

**Error**

```
Error: Command failed: powershell.exe -NoProfile -NonInteractive -InputFormat None -Command Get-AuthenticodeSignature -LiteralPath 'C:\Users\sgtpo\AppData\Local\ipfs-desktop-updater\pending\temp-IPFS-Desktop-Setup-0.30.2.exe' | ConvertTo-Json -Compress | ForEach-Object { [Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($_)) }
Get-AuthenticodeSignature : File 
C:\Users\sgtpo\AppData\Local\ipfs-desktop-updater\pending\temp-IPFS-Desktop-Setup-0.30.2.exe was not found.
At line:1 char:1
  Get-AuthenticodeSignature -LiteralPath 'C:\Users\sgtpo\AppData\Local\ ...
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      CategoryInfo          : ObjectNotFound: (:) [Get-AuthenticodeSignature], FileNotFoundException
      FullyQualifiedErrorId : SignatureCommandsBaseFileNotFound,Microsoft.PowerShell.Commands.GetAuthenticodeSignature 
   Command
 

    at ChildProcess.exithandler (node:child_process:408:12)
    at ChildProcess.emit (node:events:526:28)
    at maybeClose (node:internal/child_process:1092:16)
    at ChildProcess._handle.onexit (node:internal/child_process:302:5)
```

FYI: I got this error after selecting "check for updates". First, I got an error saying "cannot download update, check internet connection and try again" and then I selected check for updates again, and got the prompt to report the error

### LINKED ISSUE #2752: [gui error report] Error: Command failed: powershell.exe -NoProfile -NonInteractive -InputF
👉️ Please describe what you were doing when this error happened.

**Specifications**

- **OS**: win32
- **IPFS Desktop Version**: 0.22.0
- **Electron Version**: 19.0.8
- **Chrome Version**: 102.0.5005.148

**Error**

```
Error: Command failed: powershell.exe -NoProfile -NonInteractive -InputFormat None -Command Get-AuthenticodeSignature -LiteralPath 'C:\Users\User 1\AppData\Local\ipfs-desktop-updater\pending\temp-IPFS-Desktop-Setup-0.33.0.exe' | ConvertTo-Json -Compress | ForEach-Object { [Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($_)) }

    at ChildProcess.exithandler (node:child_process:408:12)
    at ChildProcess.emit (node:events:526:28)
    at maybeClose (node:internal/child_process:1092:16)
    at ChildProcess._handle.onexit (node:internal/child_process:302:5)
```

### LINKED ISSUE #2778: [gui error report] Error: Command failed: powershell.exe -NoProfile -NonInteractive -InputF
👉️ Please describe what you were doing when this error happened.

**Specifications**

- **OS**: win32 10.0.19045
- **IPFS Desktop Version**: 0.33.0
- **Electron Version**: 19.1.9
- **Chrome Version**: 102.0.5005.167

**Error**

```
Error: Command failed: powershell.exe -NoProfile -NonInteractive -InputFormat None -Command Get-AuthenticodeSignature -LiteralPath 'C:\Users\بيسسسس\AppData\Local\ipfs-desktop-updater\pending\temp-IPFS-Desktop-Setup-0.34.0.exe' | ConvertTo-Json -Compress | ForEach-Object { [Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($_)) }

    at ChildProcess.exithandler (node:child_process:408:12)
    at ChildProcess.emit (node:events:526:28)
    at maybeClose (node:internal/child_process:1092:16)
    at ChildProcess._handle.onexit (node:internal/child_process:302:5)
```

### LINKED ISSUE #2779: [gui error report] Error: Command failed: powershell.exe -NoProfile -NonInteractive -InputF
👉️ Please describe what you were doing when this error happened.

**Specifications**

- **OS**: win32 10.0.19045
- **IPFS Desktop Version**: 0.33.0
- **Electron Version**: 19.1.9
- **Chrome Version**: 102.0.5005.167

**Error**

```
Error: Command failed: powershell.exe -NoProfile -NonInteractive -InputFormat None -Command Get-AuthenticodeSignature -LiteralPath 'C:\Users\jesus\AppData\Local\ipfs-desktop-updater\pending\temp-IPFS-Desktop-Setup-0.34.0.exe' | ConvertTo-Json -Compress | ForEach-Object { [Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($_)) }

    at ChildProcess.exithandler (node:child_process:408:12)
    at ChildProcess.emit (node:events:526:28)
    at maybeClose (node:internal/child_process:1092:16)
    at ChildProcess._handle.onexit (node:internal/child_process:302:5)
```

### LINKED ISSUE #2780: [gui error report] Error: Command failed: powershell.exe -NoProfile -NonInteractive -InputF
👉️ Please describe what you were doing when this error happened.

**Specifications**

- **OS**: win32 10.0.19045
- **IPFS Desktop Version**: 0.33.0
- **Electron Version**: 19.1.9
- **Chrome Version**: 102.0.5005.167

**Error**

```
Error: Command failed: powershell.exe -NoProfile -NonInteractive -InputFormat None -Command Get-AuthenticodeSignature -LiteralPath 'C:\Users\ErickWatson\AppData\Local\ipfs-desktop-updater\pending\temp-IPFS-Desktop-Setup-0.34.0.exe' | ConvertTo-Json -Compress | ForEach-Object { [Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($_)) }

    at ChildProcess.exithandler (node:child_process:408:12)
    at ChildProcess.emit (node:events:526:28)
    at maybeClose (node:internal/child_process:1092:16)
    at ChildProcess._handle.onexit (node:internal/child_process:302:5)
```

### LINKED ISSUE #2784: [gui error report] Error: Command failed: powershell.exe -NoProfile -NonInteractive -InputF
👉️ Please describe what you were doing when this error happened.

**Specifications**

- **OS**: win32 10.0.22621
- **IPFS Desktop Version**: 0.27.0
- **Electron Version**: 19.1.9
- **Chrome Version**: 102.0.5005.167

**Error**

```
Error: Command failed: powershell.exe -NoProfile -NonInteractive -InputFormat None -Command Get-AuthenticodeSignature -LiteralPath 'C:\Users\jrahl\AppData\Local\ipfs-desktop-updater\pending\temp-IPFS-Desktop-Setup-0.34.0.exe' | ConvertTo-Json -Compress | ForEach-Object { [Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes($_)) }

    at ChildProcess.exithandler (node:child_process:408:12)
    at ChildProcess.emit (node:events:526:28)
    at maybeClose (node:internal/child_process:1092:16)
    at ChildProcess._handle.onexit (node:internal/child_process:302:5)
```

### LINKED ISSUE #2813: Restore macOS builds to use ARM-based `macos-latest`
This is the actual fix that needs to happen after workaround from https://github.com/ipfs/ipfs-desktop/pull/2811 was merged.

The way we build for macOS right now is to do Intel build, and leverage intel emulation on ARM.
This works, we have single build that works on both ARM and Intel, because intel Kubo binary is also used.

Unfortunately it is not future-proof, as ARM is the bew default.

Regressions like https://github.com/ipfs/ipfs-desktop/issues/2809 will happen. 
Right now we are not able to build on `macos-latest` (14) because it is ARM-based, and there is no Intel version anymore. 

## Next steps

We need to rework the way our macOS build works.

There is some prior work and research done in https://github.com/ipfs/ipfs-desktop/pull/1856 and in https://github.com/ipfs/ipfs-desktop/issues/2681. 

Initial thoughts:
- Prioritize ease of maintenance. 
- Producing universal DMG sounds like a way to go, as long we have extra code that ensures  correct Kubo binary is used
  - This could be done at runtime, or we could fetch both binaries (intel and arm) at build time, and manually merge them into a universal one. This will work, because the new kubo binary will be notarized as part of DMG.

### PR DESCRIPTION
**Pull request by lidel**
This is exploration into better support of Apple Silicon.

[electron-builder](https://github.com/electron-userland/electron-builder) supports producing either universal package or one for each architecture. 

In case of Desktop, things are more complicated: we also bundle go-ipfs binary. 
This means that unless the go-ipfs binary  is also universal or per-arch, we can't improve much.

Due to the way we depends on  go-ipfs via  [npm-go-ipfs](https://github.com/ipfs/npm-go-ipfs) universal binaries all the way sound like way easier route (downside being the disk size), however dist.ipfs.io already has  amd64 and arm64  for darwin.

### Universal binary test

Main cost of universal package is size, even with out go-ipfs (still being amd64) it adds ~75 MB:

- 195M   IPFS Desktop-0.16.0-universal.dmg (electron amd64&arm64 + go-ipfs arm64)
- 119 MB IPFS-Desktop-0.16.0.dmg (amd64, M1 will use Roseta)

##  TODO

- [x] decide if we want to provide single universal binary or one for amd64 (legacy intel)  and one for arm64 (new apple silicon)
  - Data point from Brave: 
    - atm IPFS integration  is using `amd64`: go-ipfs binary is amd64, it runs on M1 via Rosetta. Brave plans on using  `arm64` in the near future, as they already provide `arm64` version of Brave itself.. In that contenxt universal binary is wasteful (IPFS component is fetched on-demand, so small size matters).
- [x] based on previous decision, decide how we handle `go-ipfs` dependency (right now it downloads `kubo` only for arch of CI worker, which is amd64). 
- [x] make Universal DMG and ZIP
- [x] replace Kubo binary with universal one
  - [x] postinstall script that detects darwin build on ARM, and fetches missing intel binary + merges both into universal and puts it at `node_modules/kubo/bin/ipfs`
- [x] upstream [kubo+0.28.0.patch](https://github.com/ipfs/ipfs-desktop/blob/88396802ab76502e6becfead817893cb7fb13270/patches/kubo%2B0.28.0.patch) to npm-kubo repo 
  - [x] opened https://github.com/ipfs/npm-kubo/pull/73 
- [x] test DMG from this PR on both 
  - [x] Intel 
  - [x] ARM: tested on M1 macMini with macOS Sonoma (14), works fine with universal Kubo binary
- [ ] ~update Electron (if needed)~ does not seem to be needed on ARM, tbd if Intel works
- [x] wait for https://github.com/ipfs/npm-kubo to ship with exposed light `download` function 
  - [x]  switch this PR to `kubo` dependency with https://github.com/ipfs/npm-kubo/pull/73 (0.29.x)
  - [x] remove `package-patch` for `kubo` dependency
- [x] fix powershell bug from https://github.com/electron-userland/electron-builder/issues/7127   
- [ ] resolve crash from https://github.com/ipfs/ipfs-desktop/pull/1856#pullrequestreview-2108685897

Closes #2813

`powershell.exe -NoProfile -NonInteractive` thing should be also fixed by https://github.com/electron-userland/electron-builder/issues/7127#issuecomment-1991876763, so:

Closes #2335
Closes #2636
Closes #2752
Closes #2778
Closes #2779
Closes #2780
Closes #2784

### CODE REVIEW

**Comment by AlexxNica** [comment]
@lidel, do we know where the difference in size comes from, specifically? We could find where they are and try something to minimize the impact of shipping the universal package.

**Comment by unbeatable-101** [comment]
> @lidel, do we know where the difference in size comes from, specifically? We could find where they are and try something to minimize the impact of shipping the universal package.

Not a dev, but I'm pretty sure the size increase is just the binaries about doubling in size due to basically containing two apps.

**Comment by jenkijo** [comment]
Hello, any update?
I've committed fix for `go-ipfs` [https://github.com/ipfs/npm-go-ipfs/pull/43](https://github.com/ipfs/npm-go-ipfs/pull/43)

**Comment by dvcrn** [comment]
Thanks for the PR! I wanted to build this locally but ran into issues with the build step trying to download a arm ipfs binary that obviously didn't exist. How did you build this?

**Comment by jenkijo** [comment]
It work fine with me, what error that you got?

**Comment by hacdias** [comment]
I think that Universal would be easier for the end users despite the storage size.

**Comment by SgtPooki** [comment]
Have we looked into https://github.com/tauri-apps/tauri at all?

**Comment by SgtPooki** [comment]
Regarding universal app vs rosetta specific.. the speed increases you get by running M1 native is significant. I don't think we should decide to just force rosetta forevermore. If the decision is to wait for some period until we have more bandwidth, I think sticking with rosetta and the smaller bundle for now is fine. Ultimately, we should be moving to the type of app bundle the platform we're targeting recommends. 

For Apple/macOS, that is the universal build.

**Comment by hacdias** [comment]
@SgtPooki an issue you may want to look at too is #867

**Comment by lidel** [comment]
> Have we looked into https://github.com/tauri-apps/tauri

I think we looked at it briefly, but it was very alpha back then.  IIRC the main downside was lack of tooling for end-to-end autoupdate solution, which is raison d'être of ipfs-desktop.  It seems they work on it, but it is still undocumented? https://github.com/tauri-apps/tauri/discussions/2776

Long term, we want to move away from fetching updates from centralized HTTP server like github and leverage IPFS+IPNS  instead (see old notes in https://github.com/ipfs/ipfs-desktop/issues/789), so if we have working setup for that, we could move away from Electron (literally the only value it brings is https://www.electron.build/auto-update working on all three platforms: macOS+Windows+Linux(AppImage))

>  75 MB won't probably affect many people.

Usually I would be against doubling the size of download, but we talk Apple here – it is fair to assume that  if someone uses mac they most likely don't care about additional download cost.

The caveat here is that we don't have "universal go-ipfs binary" – iiuc this PR (in current form) still bundles the intel version of go-ipfs.  

Figuring out how to include both binaries  (arm64 and amd64) for darwin  is the key to unblocking this.

**Comment by Whitedragon2772** [comment]
lo

**Comment by SgtPooki** [comment]
> I wrote a plan for unblocking this in https://github.com/ipfs/distributions/issues/770#issuecomment-1329178981.
> 
> Not a macOS user myself, but if someone wants to work on this, I'm around to advise.

I read that comment and the approach seems reasonable and clear. Thanks for the backlink!

**Comment by lidel** [comment]
There are two separate topics at play (mind I am not macOS user, below is my rough understanding on Apple Silicon situation)

- architecture of go binary (kubo daemon) 
  -  What we did on dist.ipfs.tech is we provide separate Kubo binaries for `amd64` (intel) and `arm64` (arm/apple silicon) (https://github.com/ipfs/distributions/pull/838, https://github.com/ipfs/distributions/pull/841). 
    - Brave downloads correct binary at runtime. 
    - There is no universal binary for Kubo, but you could build one as noted in https://github.com/ipfs/distributions/issues/770#issuecomment-1329178981 – afaik nobody is doing that atm.
      - IPFS Desktop DMG bundles NPM version pulled in during the build – afaik this is still amd64, so it runs emulation.
      - IPFS Desktop needs to either stop embeedding and fetch correct binary at runtime, or include both and run correct one on macOS. 
- architecture of electron app (ipfs desktop)
  - this feels unresolved, but maybe not a big deal, since the majority of work happens in kubo?
  -  iiuc what happens on macOS is that we produce DMG that is amd64 and when run on Apple Silicon it runs emulation


I agree we should close this, and if we want to use native arch on macOS, need to fill new issue and make a new plan for both topics above.

**Comment by SgtPooki** [comment]
please see https://github.com/ipfs/ipfs-desktop/issues/2681 for a continuation of this work

**Comment by lidel** [comment]
Reopening because of https://github.com/ipfs/ipfs-desktop/issues/2813

I will prototype Universal DMG that includes Universal Kubo binary as well.

**Comment by lidel** [comment]
Fetching both `x86_64` `arm64` and merging them into universal Kubo binary is working ([build log](https://github.com/ipfs/ipfs-desktop/actions/runs/9119544018/job/25075202983#step:9:54)):

`→ lipo info after merging binaries
Architectures in the fat file: /Users/runner/work/ipfs-desktop/ipfs-desktop/node_modules/kubo/kubo/ipfs are: x86_64 arm64` 

Remaining work is to upstream [kubo+0.28.0.patch](https://github.com/ipfs/ipfs-desktop/blob/88396802ab76502e6becfead817893cb7fb13270/patches/kubo%2B0.28.0.patch) to [npm-kubo](https://github.com/ipfs/npm-kubo) repo and test DMG from this PR on both Intel and ARM macOS.

**Comment by 2color** [comment]
Also tested Tested https://dweb.link/ipfs/bafybeidw62sobej3qlfsmxxw5bbnczz6ithimvixf3xg7rwquur3iww66q/ipfs-desktop-0.35.1-mac.dmg on ARM and it worked. Can test on a MacOS Intel tonight.

**Review by SgtPooki — CHANGES_REQUESTED** [review]
Should this be version `0.36.0`? code changes look fine. app shows universal. but I tried running the app ([from latest CI job](https://github.com/ipfs/ipfs-desktop/actions/runs/9454523632?pr=1856)) and it crashes immediately.

image


```
System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGBUS)
Exception Codes:       KERN_PROTECTION_FAILURE at 0x0000007483f72aa9
Exception Codes:       0x0000000000000002, 0x0000007483f72aa9

Termination Reason:    Namespace SIGNAL, Code 10 Bus error: 10
Terminating Process:   exc handler [24263]

VM Region Info: 0x7483f72aa9 is in 0x74005c0000-0x7500000000;  bytes after start: 2207984297  bytes before end: 2080953686
      REGION TYPE                    START - END         [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      Memory Tag 255             7400540000-74005c0000   [  512K] rw-/rwx SM=ZER  
--->  Memory Tag 255             74005c0000-7500000000   [  4.0G] ---/rwx SM=ZER  
      Memory Tag 255             7500000000-7500004000   [   16K] rw-/rwx SM=ZER  
```

**Comment by lidel** [comment]
@SgtPooki yes, this will be released as 0.36.0 (2e891f3). 

The crash is concerning. 
- Are you on Intel or Apple Silicon machine?
- Is it crashing every time? 
- Was it crashing with RC 2 ([.zip of build here](https://github.com/ipfs/ipfs-desktop/actions/runs/9421878205/artifacts/1580337277))?

**Comment by SgtPooki** [comment]
> * Are you on Intel or Apple Silicon machine?

Apple silicon

> * What is the version of macOS?

Sonoma 14.5 (23F79)

> * Is it crashing every time?

Yep.

> * Was it crashing with RC 2 ([.zip of build here](https://github.com/ipfs/ipfs-desktop/actions/runs/9421878205/artifacts/1580337277))?
> 
> I can get my hands on M1 macMini tomorrow, but having second data point would be useful.

that version is also crashing for me immediately. I remember a previous version you asked me to test a month ago or so worked fine. If that's the same one, it's not now and the problem is my system.

**Review by SgtPooki — APPROVED** [review]

File: package.json
Comment (lidel) [inline package.json:81]: :information_source:  This includes fix for `powershell.exe -NoProfile -NonInteractive` errors:
https://github.com/electron-userland/electron-builder/issues/7127#issuecomment-1991876763

**Comment by lidel** [comment]
Thanks, sounds like it.

FWIW I've just tested both on M1 macMini and both work ok. Let's merge and ship.

If crashes of IPFS Desktop binary (electron) rather than kubo itself become common, then the first thing we should do is to upgrade to supported version of Electron (#2775)

**Pull request merged**
